Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp

```diff
@@ -7,390 +7,244 @@ For license and copyright information please follow this link:
+		} else if (majorVersion == 1 && minorVersion) {
+			imageKey = "image_data";
+		} else if ((majorVersion == 1 && minorVersion < 1) || majorVersion < 1) {
+			imageKey = "icon_data";
+		} else {
+			LOG(("Native notification error: unknown specification version"));
+			return;
 		}
+	} else {
+		LOG(("Native notification error: specification version is null"));
+		return;
 	}
-#endif // !TDESKTOP_DISABLE_GTK_INTEGRATION
-}
 
-#ifndef TDESKTOP_DISABLE_GTK_INTEGRATION
-class Manager::Private {
-public:
-	using Type = Window::Notifications::CachedUserpics::Type;
-	explicit Private(Type type)
-	: _cachedUserpics(type) {
-	}
+	auto image = QImage(imagePath).convertToFormat(QImage::Format_RGBA8888);
+	QByteArray imageBytes((const char*)image.constBits(), image.sizeInBytes());
 
-	void init(Manager *manager);
+	ImageData imageData;
+	imageData.width = image.width();
+	imageData.height = image.height();
+	imageData.rowStride = image.bytesPerLine();
+	imageData.hasAlpha = true;
+	imageData.bitsPerSample = 8;
+	imageData.channels = 4;
+	imageData.data = imageBytes;
 
-	void showNotification(
-		not_null<PeerData*> peer,
-		MsgId msgId,
-		const QString &title,
-		const QString &subtitle,
-		const QString &msg,
-		bool hideNameAndPhoto,
-		bool hideReplyButton);
-	void clearAll();
-	void clearFromHistory(not_null<History*> history);
-	void clearNotification(PeerId peerId, MsgId msgId);
+	_hints[imageKey] = QVariant::fromValue(imageData);
+}
 
-	bool hasPoorSupport() const {
-		return _poorSupported;
-	}
-	bool hasActionsSupport() const {
-		return _actionsSupported;
+std::vector<QString> NotificationData::getServerInformation() {
+	std::vector<QString> serverInformation;
+	auto serverInformationReply = _notificationInterface.call("GetServerInformation");
+
+	if (serverInformationReply.type() == QDBusMessage::ReplyMessage) {
+		for (const auto &arg : serverInformationReply.arguments()) {
+			if (static_cast<QMetaType::Type>(arg.type()) == QMetaType::QString) {
+				serverInformation.push_back(arg.toString());
+			} else {
+				LOG(("Native notification error: all elements in GetServerInformation should be strings"));
+			}
+		}
+	} else if (serverInformationReply.type() == QDBusMessage::ErrorMessage) {
+		LOG(("Native notification error: %1").arg(QDBusError(serverInformationReply).message()));
+	} else {
+		LOG(("Native notification error: error while getting information about notification daemon"));
 	}
 
-	~Private();
+	return serverInformation;
+}
 
-private:
-	QString escapeNotificationText(const QString &text) const;
-	void showNextNotification();
+std::vector<QString> NotificationData::getCapabilities() {
+	QDBusReply<QStringList> capabilitiesReply = _notificationInterface.call("GetCapabilities");
 
-	struct QueuedNotification {
-		PeerData *peer = nullptr;
-		MsgId msgId = 0;
-		QString title;
-		QString body;
-		bool hideNameAndPhoto = false;
-	};
+	if (capabilitiesReply.isValid()) {
+		return capabilitiesReply.value().toVector().toStdVector();
```

Comment:
In my opinion, we can remove the blank elements from the vector right away.

```suggestion
		auto capabilities = capabilitiesReply.value();
		capabilities.removeAll(QString());
		return capabilities | ranges::to_vector;
```

Reply:
Which blank elements? o_O

Reply:
You have the following lines in your code.

https://github.com/telegramdesktop/tdesktop/blob/3a0c597682890ab15b3b2c61a02adf09093c989f/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp#L43-L45

So you are probably assuming that there may be blank elements?

Reply:
No, it is a list of daemon capabilities for debug purposes.
It is also a port of https://github.com/telegramdesktop/tdesktop/blob/0c0c8f3584b500ef053bb1a4d9165cdbd9e18394/Telegram/SourceFiles/platform/linux/notifications_manager_linux.cpp#L371

Reply:
Oops, I misread the condition, sorry.